File: AstronomyObjC/AstronomyObjC/OperationQueue/FetchPhotoOperation.m
//
//  FetchPhotoOperation.m
//  AstronomyObjC
//
//  Created by Nick Nguyen on 4/23/20.
//  Copyright © 2020 Nick Nguyen. All rights reserved.
//

#import "FetchPhotoOperation.h"
#import "AstronomyObjC-Swift.h"
#import "Sol.h"

@interface FetchPhotoOperation ()

@property (nonatomic, strong, readwrite, nullable) UIImage *image;
@property (nonatomic, strong) NSURLSessionDataTask *dataTask;
@property BOOL internalIsExecuting;
@property BOOL internalIsFinished;


@end

@implementation FetchPhotoOperation {


    BOOL _isFinished;
    BOOL _isExecuting;
}
//MARK:- Properties 
- (BOOL)isAsynchronous { return YES; }
- (BOOL)isConcurrent { return YES; }
+ (NSSet *)keyPathsForValuesAffectingInternalIsExecuting { return [NSSet setWithObject:@"internalIsExecuting"] ;}
- (BOOL)isExecuting { return self.internalIsExecuting; }
+ (NSSet *)keyPathsForValuesAffectingInternalIsFinished {
    return [NSSet setWithObject:@"internalIsFinished"];
}
- (BOOL)isFinished { return self.internalIsFinished; }

- (instancetype)initWithPhotoReference:(Sol *)solReference {
    self = [super init];
    if (self) {
        _solReference = solReference;
    }
    return self;
}


- (void)start {
    self.internalIsExecuting = YES;
    NSURL * url = [NSURL URLWithString:self.solReference.imageURL].usingHTTPS;
   
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *task = [session dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (self.isCancelled) {
            self.internalIsFinished = YES;
            self.internalIsExecuting = NO;
            return;
        }
        if (error) {
            NSLog(@"No data %@",error);
        }
        
        if (data) {
            self.image = [UIImage imageWithData:data];
        }
        self.internalIsExecuting = NO;
        self.internalIsFinished = YES;
        
    }];
    [task resume];
    self.dataTask = task;
    
}

-(void)cancel
{
    [self.dataTask cancel];
    [super cancel];
}














@end

File: CARAstronomy/CARAstronomy/Support/Operations/FetchPhotoOperation.swift
//
//  FetchPhotoOperation.swift
//  CARAstronomy
//
//  Created by Chad Rutherford on 2/25/20.
//  Copyright © 2020 Chad Rutherford. All rights reserved.
//

import Foundation

class FetchPhotoOperation: ConcurrentOperation {
    
    var sol: MarsSol
    var image: UIImage?
    
    init(sol: MarsSol) {
        self.sol = sol
    }
    
    override func start() {
        self.state = .isExecuting
        
        guard let url = URL(string: sol.imageURL)?.usingHTTPS else { return }
        URLSession.shared.dataTask(with: url) { data, response, error in
            defer { self.state = .isFinished }
            guard error == nil else { return }
            guard let response = response as? HTTPURLResponse, response.statusCode == 200 else { return }
            guard let data = data else { return }
            guard let image = UIImage(data: data) else { return }
            self.image = image
        }.resume()
    }
}
